admin: add connectrpc and /v2 routes

rockwotj opened the pull request:
Builds upon https://github.com/redpanda-data/redpanda/pull/26468 to plugin the protobuf generator as well as the generated RPC services
to the admin API.

These are available under `/v2` paths for our admin API. Right now
there are two services: AdminService and DebugService.


*AdminService*: a very simple service that hosts some "reflection" API endpoints like what routes are available via `/v2` in the server.

*DebugService*: A replication of some of our debug endpoints, although I think some of those should probably get broken up into different services (maybe the stress fiber thing should go into a finjector service).

The DebugService is in the `internal` package, which means it has no compatibility rules enforced and it's for Redpanda usage only.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

rockwotj committed:
admin: add support for registering v2 services to the admin server

rockwotj committed:
pbgen: remove version from RPCs

This was not a good idea, with backports, it is not actually easy to
determine the version an RPC will be introduced, so a better approach to
discovering what is supported is to just have an endpoint report all
RPCs that are supported.

rockwotj committed:
pbgen: add RPC name to route metadata

This helps with adding debug info and will be used in the get route
"reflection" request in the admin server.

Additionally, bring our default version string in alignment with the
version string when there is a real version present (adding a v prefix)

rockwotj committed:
admin: add initial admin service into the admin server

rockwotj commented:
```shell
curl localhost:9644/v2/redpanda.core.admin.AdminService/GetRoutes \
  -X POST -H 'Content-Type: application/json' \
  -d '{}' \
  | jq
{
  "routes": [
    {
      "name": "redpanda.core.admin.AdminService.GetVersion",
      "http_route": "/v2/redpanda.core.admin.AdminService/GetVersion"
    },
    {
      "name": "redpanda.core.admin.AdminService.GetRoutes",
      "http_route": "/v2/redpanda.core.admin.AdminService/GetRoutes"
    }
  ]
}
```

rockwotj committed:
admin: add initial internal services

Internal services have 0 public facing guarentees and are really for us
only.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--73b5ee0e-2fdd-4cc8-8541-6853b6cfbb71-->
<!--TEST RESULTS-->
<!--95c1ccc8-6f20-4faf-9d6b-d0691b0bdecf-->
<!--TEST RESULTS-->
<!--b623f7cb-1c67-4345-ba1c-929a837603b8-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69369">build#69369</a></summary>


| test_class       | test_method              | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------|:-------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeE2ETests | test_json_schema_unicode | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69369#01982e24-f575-495f-ae63-dafa56679797 | FLAKY         | 19/21    | upstream reliability is '98.45814977973568'. current run reliability is '90.47619047619048'. drift is 7.98196 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69441">build#69441</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69441#019832d0-c4b9-41ef-92d7-e66e3c510043 | FLAKY         | 19/21    | upstream reliability is '96.60297239915074'. current run reliability is '90.47619047619048'. drift is 6.12678 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69776">build#69776</a></summary>


| test_class       | test_method       | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 |
|:-----------------|:------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|
| AdminApiAuthTest | test_scram_sha512 | null             | integration | https://buildkite.com/redpanda/redpanda/builds/69776#01985223-00c6-4702-9b79-baf6c569d689 | FAIL          | 0/21     | The test has failed across all retries |
| AdminApiAuthTest | test_scram_sha512 | null             | integration | https://buildkite.com/redpanda/redpanda/builds/69776#01985226-8889-4ba5-b23f-827af3b19495 | FAIL          | 0/21     | The test has failed across all retries |
</details>
<!--/body-->

mmaslankaprv commented on src/v/redpanda/admin/server.cc:
```diff
@@ -340,6 +341,33 @@ admin_server::admin_server(
 
+void admin_server::add_service(
+  std::unique_ptr<serde::pb::rpc::base_service> service) {
+    for (auto& route : service->all_routes()) {
```
nit: maybe we can emit a log line when registering a route ?

rockwotj replied:
Done in https://github.com/redpanda-data/redpanda/pull/26920/commits/1b1edf94f9084f7a0b1357e2f1d2855962ab7aff

rockwotj committed:
admin: log routes and rpc services that are added

mmaslankaprv commented:
This looks great, i think the only last thing missing is a very simple ducktape test verifying if we can list routes with the newly added endpoint

michael-redpanda commented on proto/redpanda/core/admin/internal/debug.proto:
```diff
@@ -0,0 +1,57 @@
+// Copyright 2025 Redpanda Data, Inc.
```
who not BSL?

rockwotj replied:
Because we want people to embed generated code from this into their apps/clients/etc

michael-redpanda replied:
ooooo ok

rockwotj committed:
proto/admin: update protos to mirror cloud patterns more closely

Additionally, support names like RPCRoute -> rpc_route in our code
generator.

rockwotj committed:
bazel: add instructions to regenerate requirements.txt

And use a requirements.in file to more simply define python packages we
use in Bazel.

rockwotj committed:
ducktape: add admin v2 test

This additionally adds generated protos to our ducktape test suite. It's
not too bad that we have to generate this (it's just a single task command),
but not quite as seemless as Bazel as they could get out of date.

rockwotj committed:
pbgen: extract out pascalToSnakeCase

For tests, because I botched this

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/69776-->
<!--d917e9d3-e5ac-485d-887a-fc4495f57ff5-->
<!--98753fd1-e8e6-4faf-9174-0ef04dfb8ad2-->
<!--title-->
## Retry command for [Build#69776](https://buildkite.com/redpanda/redpanda/builds/69776)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/admin_api_auth_test.py::AdminApiAuthTest.test_scram_sha512
```
<!--/body-->

dotnwat commented on proto/redpanda/core/admin/admin.proto:
```diff
@@ -16,25 +16,50 @@ syntax = "proto3";
+}
+
+message GetRoutesRequest {}
+message GetRoutesResponse {
```
>so a better approach to
discovering what is supported is to just have an endpoint report all
RPCs that are supported.

+1

dotnwat commented on src/v/redpanda/admin/services/admin.cc:
```diff
@@ -0,0 +1,48 @@
+admin::admin_service_impl::get_version(proto::get_version_request) {
+    proto::get_version_response resp;
+    resp.set_version(ss::sstring(redpanda_git_version()));
+    resp.set_build_sha(ss::sstring(redpanda_git_revision()));
```
did you intend to shove different strings into these?

rockwotj replied:
they are different strings? `git_version` vs `git_revision`

dotnwat replied:
i see yeh that comment was pre-nap

rockwotj committed:
pbgen: simplify codegen for rpc handlers

dotnwat commented on bazel/thirdparty/requirements.in:
```diff
@@ -0,0 +1,8 @@
+# List all our python requirements here, one per line.
+# These can then be locked with a tool like `uv`:
+# uv pip compile --generate-hashes requirements.in -o requirements.txt
```
😎

rockwotj replied:
For the life of me I couldn't figure out how to generate the hashes :)

dotnwat commented on tests/rptest/tests/admin_api_auth_test.py:
```diff
@@ -107,6 +108,16 @@ def test_scram_sha512(self):
+            admin_v2 = AdminV2(self.redpanda,
+                               auth=(charles.username, charles.password),
+                               protocol=protocol)
+            build_info = admin_v2.get_build_info(
+                admin_pb.GetBuildInfoRequest())
+            self.logger.info(
+                f"Build info: version={build_info.version}, sha={build_info.build_sha}"
```
I guess this will implicitly test that the RPC didn't have an HTTP error, decoded correctly, etc..?

rockwotj replied:
yes exactly. I'm working on adding the formal generated python endpoints to ducktape, but requires upgrading a bunch of stuff. You can follow that here: https://github.com/redpanda-data/redpanda/pull/27002

Then I can use https://github.com/connectrpc/connect-python to generate a client for ducktape (see https://github.com/redpanda-data/vtools/pull/3780)

rockwotj replied:
once I have a real client I will expand the test surface here more.
